enable hydra jobs for packages x86_64-linux does not support

matthewbauer opened the pull request:
###### Motivation for this change
Right now, Hydra will only list packages that actually support x86_64-linux. This allows "broken" packages on x86_64-linux to be listed for the purpose of having Darwin-only packages built, available, and listed on Hydra.

This was based on @dezgeg's commits and rebased onto latest staging. I'm not sure how Hydra works in this regard, but hopefully the missing packages will show up in staging as soon as it is merged.

/cc @dezgeg @LnL7 @Ericson2314 

related to #25930
Fixes #25200

dezgeg committed:
nixpkgs release: Fix Darwin-only jobs

Currently the logic of generating nixpkgs Hydra jobs is to walk through
the pkgs evaluated for system = "x86_64-linux", collect any derivations
and their meta.platforms values. However, that doesn't work for
packages whose meta.platforms doesn't include x86_64-linux, as just
evaluating their meta attribute raises an error so they get skipped
completely.

As a less-intrusive fix (i.e. anything than rewriting the current package
enumeration logic), allow passing `config.allowUnsupportedSystem = true`
to permit evaluating packages regardless of their platform and use that
in the package listing phase.

Fixes #25200

dezgeg committed:
darwin: Add recurseForDerivations

FRidh commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -40,6 +40,8 @@ let
 
   allowBroken = config.allowBroken or false || builtins.getEnv "NIXPKGS_ALLOW_BROKEN" == "1";
 
+  allowUnsupportedSystem = allowBroken || config.allowUnsupportedSystem or false;
```
nitpick: `allowBroken` in my opinion should be kept out of here and be explicitly added in line 182 because brokenness and whether a system is supported are two different things.

Ericson2314 reviewed (approved):
Wonderful! But please also remove my dirty hack https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/release.nix#L129-L131 :)

LnL7 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11763,7 +11763,7 @@ with pkgs;
   darwin = let
     apple-source-releases = callPackage ../os-specific/darwin/apple-source-releases { };
-  in apple-source-releases // rec {
+  in recurseIntoAttrs (apple-source-releases // rec {
```
Not sure if we want to build everything in here, there's some stuff in there that depends on xcode and I don't know if all of the licences are correct.

matthewbauer replied:
Shouldn't it be marked with unfree license in that case?

matthewbauer replied:
apple-source-releases should be okay to build, right? Maybe we can just use that?

Questionable ones might be:
- usr-include
- osx_private_sdk
- xcode
- cf-private

(they are probably already being built just not listed on hydra because they are dependencies of different things)

LnL7 replied:
We could explicitly disable those with `cf-private = {};` in the release.nix, apart from that it should be ok.

matthewbauer replied:
it should be ```darwin.cf-private = {}```, right?

LnL7 replied:
yeah, I think that should work.

LnL7 replied:
apparently not, but something like this looks ok.

```nix
      darwin = packagePlatforms pkgs.darwin // {
        cf-private = {};
        osx_private_sdk = {};
        xcode = {};
      };

```

dezgeg replied:
That doesn't prevent those packages from getting built & redistributed via indirect dependencies. E.g. `vim_configurable` depends on `cf-private`, so that just looks like a no-op here.

matthewbauer replied:
Well that's how it's always been actually. This PR didn't change any of that- just what is listed by Hydra changes.

matthewbauer committed:
nixpkgs release: remove macOS hack

was needed for testing previously
see:

https://github.com/NixOS/nixpkgs/pull/28174#pullrequestreview-55944725

matthewbauer commented:
Any objections to merging?

(I don't actually have permissions to merge. If anyone wants to add me, it would be greatly appreciated!)

Ericson2314 commented:
If https://github.com/NixOS/nixpkgs/pull/28174#discussion_r132821409 is taken care of, and @LnL7 doesn't beat me to it, I'll merge tomorrow.

matthewbauer committed:
release.nix: disable private apple sdks in release

we shouldn’t let it get into the release. This includes:

- darwin.cf-private
- darwin.osx_private_sdk
- xcode
